src/frontend/src/optimizer/plan_node/logical_vector_search.rs

```diff
@@ -0,0 +1,338 @@
+                    required_cols_new_output_mapping[new_output_idx]
+                        .replace(non_distance_required_input_idx.len()),
+                    None
+                );
+                non_distance_required_input_idx.push(*input_idx);
+                required_input_idx_bitset.set(*input_idx, true);
+            }
+        }
+        if let Some(distance_new_output_idx) =
+            required_cols
+                .iter()
+                .position(|required_original_output_idx| {
+                    *required_original_output_idx == self.core.output_input_idx.len()
+                })
+        {
+            assert_eq!(
+                required_cols_new_output_mapping[distance_new_output_idx]
+                    .replace(non_distance_required_input_idx.len()),
+                None
+            );
+        }
+        let required_cols_new_output_idx = required_cols_new_output_mapping
+            .into_iter()
+            .map(Option::unwrap)
+            .collect_vec();
+        let input_required_idx = required_input_idx_bitset.ones().collect_vec();
+
+        let new_input = self.input().prune_col(&input_required_idx, ctx);
+        // mapping from idx of original input to new input
+        let mut mapping = ColIndexMapping::with_remaining_columns(
+            &input_required_idx,
+            self.input().schema().len(),
+        );
+
+        let mut new_core = self.core.clone_with_input(new_input);
+        new_core.left = mapping.rewrite_expr(new_core.left);
+        new_core.right = mapping.rewrite_expr(new_core.right);
+        new_core.output_input_idx = non_distance_required_input_idx
+            .iter()
+            .map(|input_idx| mapping.map(*input_idx))
+            .collect();
+        let vector_search = Self::with_core(new_core);
+        let plan: PlanRef = vector_search.into();
+        if required_cols_new_output_idx.len() == plan.schema().len()
+            && required_cols_new_output_idx.is_sorted()
+        {
+            // the current plan output has match the required column order.
+            plan
+        } else {
+            LogicalProject::with_out_col_idx(plan, required_cols_new_output_idx.iter().copied())
+                .into()
+        }
+    }
+}
+
+impl ExprRewritable<Logical> for LogicalVectorSearch {
+    fn has_rewritable_expr(&self) -> bool {
+        true
+    }
+
+    fn rewrite_exprs(&self, r: &mut dyn ExprRewriter) -> PlanRef {
+        let mut core = self.core.clone();
+        core.rewrite_exprs(r);
+        Self::with_core(core).into()
+    }
+}
+
+impl ExprVisitable for LogicalVectorSearch {
+    fn visit_exprs(&self, v: &mut dyn ExprVisitor) {
+        self.core.visit_exprs(v);
+    }
+}
+
+impl PredicatePushdown for LogicalVectorSearch {
+    fn predicate_pushdown(
+        &self,
+        predicate: Condition,
+        ctx: &mut PredicatePushdownContext,
+    ) -> PlanRef {
+        gen_filter_and_pushdown(self, predicate, Condition::true_cond(), ctx)
```

Comment:
Considering we have an `output_indices`, we need to apply the o2i_mapping to the predicate before pushing the predicate to the input.